Final PR for version 1.0

Kavignon opened the pull request:
For a presentable MVP for december the 9th, here are the needed functionalities to be completed 

- The experience system

- The inventory system 

- The enhanced brain 

- The link between the levels

- The shop must allow players to buy and sell items 

-  The player can't overweight the inventory

- Players can level up their characters when the requirements are met (number of experience 

- points)

- The players can set the difficulty to either Easy (random choices), Moderate (AI from augmented tactics which selects the best set of commands to execute from 3-4 scenarios), Hard (Advanced brain making it's own decision based on the training) or Lunatic (Enhanced brain making its own decision based on the training).

To be presentable, of course, a 2D prototype must be done via either Unity or Urho3D

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/gameoflightandshadows/a-game-of-light-and-shadows/97)
<!-- Reviewable:end -->

Commit:
#95 - Completed

Commit:
#95 - Completed

Commit:
The experience system manager will firstly compute the gain for the character which has engaged an enemy or was attacked by an enemy. Then, if enough experience points were accumulated, levelup points will be attributed and the experience points will be set to 0.

Commit:
#42 - Implemented the inventory items that players will be able to find in the game

review-ninja commented:
<a href="https://app.review.ninja/GameOfLightAndShadows/A-game-of-light-and-shadows/pull/97" target="_blank"><img src="https://app.review.ninja/66727011/pull/97/badge" alt="ReviewNinja"/>

Commit:
Added the InventoryItem abstract class which will be used by items instead of defining items with a parameter
Define the inventory as an array of InventoryItem because the array will be fast enough to perform the required tasks
Added the relevant function to sort properly the list so the player can see the items sorted from lightest to heaviest or from heaviest to lightest #96
The record for the inventory needs to be immuable and be changed only by the inventory manager when it has to update its state.

Kavignon commented:
Problem with actual branch (FirstPrototype).. Will merge and make another PR.
